Tasks/PublishTestResults/publishresultsthroughexe.ts

```diff
@@ -0,0 +1,112 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as tl from 'vsts-task-lib/task';
+import * as tr from 'vsts-task-lib/toolrunner';
+import * as vsts from 'vso-node-api';
+
+export class TestResultsPublisher {
+    constructor(matchingTestResultsFiles: string[], mergeResults: string, platform: string, config: string,
+        testRunTitle: string, publishRunAttachments: string, testRunner: string) {
+        
+        this.matchingTestResultsFiles = matchingTestResultsFiles.slice(0);
+        this.mergeResults = mergeResults;
+        this.platform = platform;
+        this.config = config;
+        this.testRunTitle = testRunTitle;
+        this.publishRunAttachments = publishRunAttachments;
+        this.testRunner = testRunner;
+    }
+
+    public async publishResultsThroughExe(): Promise<number> {
+        let testResultsPublisherTool: tr.ToolRunner = tl.tool(this.getTestResultsPublisherLocation());
+        let envVars: { [key: string]: string; } = this.getEnvironmentVariables();
+        let args: string[] = this.getArguments();
+        testResultsPublisherTool.arg(args);
+        
+        let exitCode: number = await testResultsPublisherTool.exec(<tr.IExecOptions>{ env: envVars, ignoreReturnCode: true });
+
+        return exitCode;
+    }
+
+    private getTestResultsPublisherLocation(): string {
+        return path.join(__dirname, 'modules/TestResultsPublisher.exe');
+    }
+
+    private getArguments(): string[] {
+        let responseFilePath = this.createResponseFile();
+        // Adding '@' because this is a response file argument
+        let args = ['@' + responseFilePath];
+        return args;
+
+    }
+
+    private createResponseFile(): string {
+        let responseFilePath: string = path.join(__dirname, 'tempResponseFile.txt');
+
+        // Adding quotes around matching file names
+        this.modifyMatchingFileName();
+        
+        // Preparing File content
+        let fileContent: string = os.EOL + this.matchingTestResultsFiles.join(os.EOL);
+
+        // Writing matching file names in the response file
+        try {
+            fs.writeFileSync(responseFilePath, fileContent);
+        }
+        catch (ex) {
+            tl.warning("Exception while writing to response file: " + ex);
```

Comment:
in case of exception aren't you supposed to return null path and fall back to agent.command?
Can we add telemetry for such failures?

Reply:
Sure, corrected it.